sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -771,6 +771,9 @@ object LikeSimplification extends Rule[LogicalPlan] {
     }
   }
 
+  private def isSimplifyMultiLike(child: Expression): Boolean =
+    child.isInstanceOf[Attribute] || child.foldable
```

Comment:
let's say if the data source can't push down string predicates, is it still useful to do multi-like optimization even if the input is an attribute?

Reply:
It may also be optimized to `StringStartsWith`, `StringEndsWith` and `StringContains`. For example:
```sql
select * from t1 where id like all('%a', 'b%', '%c%', '%a%b%', '%a%b%c%')
```
```
== Physical Plan ==
*(1) Filter ((((isnotnull(id#7) AND EndsWith(id#7, a)) AND StartsWith(id#7, b)) AND Contains(id#7, c)) AND likeall(id#7, %a%b%, %a%b%c%))
+- *(1) ColumnarToRow
   +- FileScan parquet spark_catalog.default.t1[id#7] Batched: true, DataFilters: [isnotnull(id#7), EndsWith(id#7, a), StartsWith(id#7, b), Contains(id#7, c), likeall(id#7, %a%b%,..., Format: Parquet, PartitionFilters: [], PushedFilters: [IsNotNull(id), StringEndsWith(id,a), StringStartsWith(id,b), StringContains(id,c)], ReadSchema: struct<id:string>

```

Reply:
OK, so multiple simple string operations is faster than a multi-like operation? Then I agree it's better to only optimize if input is a cheap expression. Can we reuse `CollapseProject.isCheap`?

Reply:
OK